field/src/extension_field/goldilocks_field.rs

```diff
@@ -0,0 +1,474 @@
+use plonky2_util::branch_hint;
+use static_assertions::const_assert;
+
+use crate::extension_field::Extendable;
+use crate::field_types::Field64;
+use crate::goldilocks_field::{reduce128, GoldilocksField};
+
+/// Reduce the value x_lo + x_hi * 2^128 to an element in the
+/// Goldilocks field.
+#[inline(always)]
+pub(crate) fn reduce160(x_lo: u128, x_hi: u32) -> GoldilocksField {
+    // If t is in {1, ..., 2^32-1},  then t*2^128 % p == p - (t << 32)
+    let hi = <GoldilocksField as Field64>::ORDER - ((x_hi as u64) << 32);
+
+    // hi is not reduced if x_hi was 0, but that doesn't matter.
+    let (lo, cy) = x_lo.overflowing_add(hi as u128);
+    if cy {
+        // cy = true is very rare. The only way it can happen is if
+        // x_lo is at least 2^128 - (2^64 - 2^63), i.e.
+        // 0xFFFFFFFF FFFFFFFF 80000000 00000000
+        // which for randomly distributed values will only happen with
+        // probability about 2^-64.
+        branch_hint();
+        let lo = reduce128(lo).0;
+        let cy_red = <GoldilocksField as Field64>::ORDER - (1u64 << 32);
+        reduce128(lo as u128 + cy_red as u128)
+    } else {
+        reduce128(lo)
+    }
```

Comment:
I think this can be made a bunch faster by implementing it in the same way as `reduce128`.

A 160-bit `x` can be split into a 64-bit `a0`, 32-bit `a1`, and a 64-bit `a2` such that `x` = `a0` + `a1` × 2^64 + `a2` × 2^96. Recall that 2^64 ≡ 2^32 − 1 (mod _P_) and 2^96 ≡ −1 (mod _P_). Therefore, `x` ≡ `a0` − `a2` + `a1` × (2^32 − 1) (mod _P_).

This is the same as `reduce128` except there `a2` is 32-bit.

Possible implementation in 7 uops (I have not tested this):
```rust
fn reduce160_small(x_lo: u128, x_hi: u32) -> GoldilocksField {
    // shld to form x_hi
    // shr to form x_mid
    let (x_lo, x_mid, x_hi) = (x_lo as u64, (x_lo >> 64) as u32, (x_lo >> 96) as u64 + ((x_hi as u64) << 32));

    // sub + jc (should fuse)
    // checked_sub includes a hint to LLVM that underflow is unlikely
    let (mut t0, borrow) = x_lo.checked_sub(x_hi);
    if borrow {
        // The maximum possible value of x is (2^64 - 1)^2 * 4 * 7 < 2 ** 133, so x_hi < 2 ** 37.
        // A borrow will happen roughly one in 134 million times, so it's best to branch.
        branch_hint();
        t0 -= EPSILON; // Cannot underflow if x_hi is canonical. // NB: this assumes that x < 2^160 - 2^128 + 2^96.
    }
    // imul
    let t1 = (x_hi_lo as u64) * EPSILON;
    // add, sbb, add
    let t2 = unsafe { add_no_canonicalize_trashing_input(t0, t1) }; // this is defined in goldilocks_field.rs
    GoldilocksField(t2)
}
```

Reply:
Nice! That improves latency by quite a lot, and improves throughput a bit too. I'll update the numbers in the PR comment for the record.

Not sure where I should put this function though. I makes some assumptions specific to the extension field multiplication (i.e. in the likelihood of a borrow), but it also calls `add_no_canonicalize_trashing_input` which is in `goldilocks_field.rs`. I currently have it in `goldilocks_field.rs` with `pub(crate)` visibility. Any preference?

Reply:
Yeah, there’s no great place to put it. I think putting it in `goldilocks_field.rs` with `pub(crate)` is fine :)